File: backend/src/models/loanProduct.js
import { DataTypes } from "sequelize";

import { sequelize } from "../db/sequelize.js";

function numericGetter(field) {
  return function get() {
    const raw = this.getDataValue(field);
    return raw === null || raw === undefined ? raw : Number(raw);
  };
}

export const LoanProduct = sequelize.define(
  "LoanProduct",
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    name: { type: DataTypes.STRING, allowNull: false, unique: true },
    description: { type: DataTypes.TEXT, allowNull: false },
    interest_rate: {
      type: DataTypes.DECIMAL(5, 2),
      allowNull: false,
      get: numericGetter("interest_rate"),
    },
    minimum_income: {
      type: DataTypes.DECIMAL(12, 2),
      allowNull: false,
      get: numericGetter("minimum_income"),
    },
    maximum_amount: {
      type: DataTypes.DECIMAL(12, 2),
      allowNull: false,
      get: numericGetter("maximum_amount"),
    },
    minimum_tenure: {
      type: DataTypes.DECIMAL(5, 0),
      allowNull: false,
      get: numericGetter("minimum_tenure"),
    },
    maximum_tenure: {
      type: DataTypes.DECIMAL(5, 0),
      allowNull: false,
      get: numericGetter("maximum_tenure"),
    },
    eligibility_rules: {
      type: DataTypes.JSON,
      allowNull: false,
      defaultValue: {},
    },
    created_at: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW,
    },
  },
  {
    tableName: "loan_products",
    timestamps: false,
  },
);

File: backend/app/models/loan_product.py
import uuid
from datetime import datetime

from sqlalchemy import JSON, Column, DateTime, Numeric, String, Text

from app.db.base import Base


class LoanProduct(Base):
    __tablename__ = "loan_products"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String, unique=True, nullable=False)
    description = Column(Text, nullable=False)
    interest_rate = Column(Numeric(5, 2), nullable=False)
    minimum_income = Column(Numeric(12, 2), nullable=False)
    maximum_amount = Column(Numeric(12, 2), nullable=False)
    minimum_tenure = Column(Numeric(5, 0), nullable=False)
    maximum_tenure = Column(Numeric(5, 0), nullable=False)
    eligibility_rules = Column(JSON, nullable=False, default=dict)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
